### DISCUSSION TITLE
videos playlist "Cannot decode file"

### DISCUSSION THREAD

**Discussion by maxguy** [savonet/liquidsoap#3648] [User Support]
I'm trying to use liquidsoap to make a web tv and then use playlists. using the "single" function the stream occurs correctly while reading the folder with the individual mp4 files always gives the message "Cannot decode file" while sending the individual files in "single" they are forwarded to the server correctly. Below is a small snipper of the code used.

mainplay = playlist('/srv/sources/VOD/main')
trailer = playlist('/srv/sources/VOD/trailer')
webtv_scheduled = switch([({0h-24h}, mainplay)])
emergency = single("/srv/sources/VOD/trailer/test.mp4")
tv_program = rotate(weights=[1,4], [trailer, webtv_scheduled])
livestream = fallback(track_sensitive=false,[tv_program,emergency])
flvenc = %ffmpeg(format = "flv", %audio.copy, %video.copy)
url = "rtmp://tech:xxxxx@10.10.10.10:1935/tvapp/webtv"
livestream = fallback(track_sensitive=false,[mainplay,emergency])
oneplayOK = single("/srv/sources/VOD/main/test.mp4";
output.url(flvenc,url=url,livestream)

**Comment by vitoyucepi** [comment]
Hi @maxguy,
I will start by refactoring your script a bit.

## Refactoring
Sources:
```ruby
mainplay = playlist("/srv/sources/VOD/main")
emergency = single("/srv/sources/VOD/trailer/test.mp4")
trailer = playlist("/srv/sources/VOD/trailer")
oneplayOK = single("/srv/sources/VOD/main/test.mp4")
```

1. ```ruby
   webtv_scheduled = switch([({0h-24h}, mainplay)])
   tv_program = rotate([trailer, webtv_scheduled], weights=[1,4])
   livestream = fallback([tv_program, emergency], track_sensitive=false)
   ```
1. ```ruby
   livestream = fallback([mainplay, emergency], track_sensitive=false)
   ```

Output:
```ruby
FLV_ENC = %ffmpeg(%audio.copy, %video.copy, format="flv")
url = "rtmp://user:password@10.10.10.10:1935/tvapp/webtv"
output.url(FLV_ENC, livestream, url=url)
```

## Questions
1. Could you please describe your setup?
   1. Which OS and version do you use?
   1. How did you install liquidsoap?
      From opam, packages or something else?
1. Could you post the error log?

**Reply by maxguy** [reply]
now i'm using a docker with debian 12 with opam enabled and liquidsoap 2.2.3dev
now something is working but i receive some errors ... i try to understand  if the problem is file encode.
i'll attach latest error logs.
can you to suggest a guide or ebook to configure correctly liquidsoap for my scope and which are best settings for playlist and rotation.
[video copy.log](https://github.com/savonet/liquidsoap/files/14001808/video.copy.log)
is it possible with video add audio optmization with filters in sync with video?
last question if you can answer is about metadata for video.. is the same way for audio?
thank you so
much

**Reply by vitoyucepi** [reply]
Tell me more about your video library.
```
2024/01/21 14:58:44 [decoder:2] Error while checking file's content: Content.Incompatible_format: formats ffmpeg.copy(codec="h264",width=854,height=480,aspect_ratio=0/1,pixel_format=yuv420p) and ffmpeg.copy(codec="h264",width=1280,height=720,aspect_ratio=0/1,pixel_format=yuv420p) are incompatible!
```

You should either change the video resolution using liquidsoap settings
```ruby
settings.frame.video.width := 854
settings.frame.video.height := 480
# Frame rate could be could also be changed
# settings.frame.video.framerate := 25
```
or use something other than copy-codec.
```ruby
# This example should be tweaked...
FLV_ENC = %ffmpeg(%audio(codec="aac"), %video(codec="libx264"), format="flv")
```

**Reply by maxguy** [reply]
Hi Vito, I made the changes as you suggested, now I'll see if at least the library doesn't have any problems. all videos have the same frame size and the same codec. Obviously something isn't right. I'll try it now and see if it goes better.
Taking advantage of your availability, I wanted to ask you how you could improve the audio of the videos, add transitions between one video and another, pass the metadata for the videos.
I'm sorry if I bother you but I can't find examples of applications that can answer my doubts. congratulations for the excellent work
Thank you

**Reply by vitoyucepi** [reply]
> I wanted to ask you how you could improve the audio of the videos

Could you please explain this request in detail?

> add transitions between one video and another

I'm not sure about video transitions, but you can take a look at page 218 of [the book](https://www.liquidsoap.info/book/book.pdf).
You might also try using ffmpeg filters.

Both methods require re-encoding the video. Copy filters won't work.

> pass the metadata for the videos.

This feature depends on the protocol and implementation, as far as I understand the whole process.

**Reply by maxguy** [reply]
thank for your response,
by audio tweaking I mean volume normalization and maybe a bit of boost equalization
for transition ok i will see the documentation of ffmpeg
by metadata I mean if it is possible to dynamically add data to the stream as reported for the guide on audio streams like the example below thanks.


`settings.log.stdout.set(true)
######################
def apply_metadata(m) =
  title = m["title"]
  artist = m["artist"]
  print("#{artist} - #{title}")
end
######################
playlist1 = playlist(mode="randomize",reload_mode="watch", "/dir1")
playlist2 = playlist(mode="randomize",reload_mode="watch", "/dir2")

radio = mksafe(rotate([playlist1, playlist2]))
radio = fallback([request.queue(id="request"),radio])

radio = audio_to_stereo(radio)
radio = on_metadata(apply_metadata, radio)`

**Reply by vitoyucepi** [reply]
> `audio_to_stereo`

The guide is a bit outdated, but the main concepts will work fine. You can read/write metadata, but the overall result will depend on your intentions.

> by audio tweaking I mean volume normalization and maybe a bit of boost equalization

If you want to normalize audio, use something like [loudgain](https://github.com/Moonbase59/loudgain) to write the gain information into the files. After that you can use the internal normalization functions. You can read about them on the https://www.liquidsoap.info/doc-dev/replay_gain.html page.

As for equalization, I think you should use ffmpeg filters.

**Reply by Moonbase59** [reply]
_loudgain_ actually supports writing ReplayGain data to some video file formats, although that’s not well documented. I would have to look it up—it’s such a long time since I wrote it…

**EDIT:** My README says:

> loudgain was never intended for video files, but it actually _does_ work on `.mp4` video files, although only one audio track is respected. Go figure!

For Matroska `.mkv`, something like [mkvrg](https://github.com/kevinlekiller/mkvrg) could be used.

Can we actually use something like

```
radio = amplify(1.,override="replaygain_track_gain",radio)
```

with video files?
